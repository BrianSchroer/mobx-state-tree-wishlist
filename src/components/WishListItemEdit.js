import React from 'react';
import { observer } from 'mobx-react';

class WishListItemEdit extends React.Component {
  render() {
    const { item } = this.props;
    return (
      <div className="item-edit">
        <div>
          <div className="item-edit-label">
            <label htmlFor="name">Thing:</label>
          </div>
          <input id="name" value={item.name} onChange={this.onNameChange} />
        </div>
        <div>
          <div className="item-edit-label">
            <label htmlFor="price">Price:</label>
          </div>
          <input id="price" value={item.price} onChange={this.onPriceChange} />
        </div>
        <div>
          <div className="item-edit-label">
            <label htmlFor="image">Image:</label>
          </div>
          <input id="image" value={item.image} onChange={this.onImageChange} />
        </div>
      </div>
    );
  }

  onNameChange = event => {
    this.props.item.changeName(event.target.value);
  }

  onPriceChange = event => {
    // this doesn't let you type decimals because it doesn't accept trailing '.'
    const price = parseFloat(event.target.value);
    if (isNaN(price)) return;

    this.props.item.changePrice(price);
  }

  onImageChange = event => {
    this.props.item.changeImage(event.target.value);
  }
}

export default observer(WishListItemEdit);