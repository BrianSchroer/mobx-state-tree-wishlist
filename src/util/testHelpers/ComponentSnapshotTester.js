import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component snapshot helper
 */
export default class ComponentSnapshotTester {

  /**
   * Create new ComponentSnapshotTester instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propOverrides) => Object.assign({}, props, propOverrides);
  }

  /**
   * Specify callback function to be called by the test function to adjust component
   * props with propOverrides.
   * (Defaults to adjusting top-level props.)
   */
  withPropAdjustor = (adjustor) => {
    this.adjustProps = adjustor;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propOverrides, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = (propOverrides) => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propOverrides);

    const newSnapshot = renderer.create(<Component {...props} />).toJSON();

    expect(newSnapshot).toMatchSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot. 
   * @param {*} component - React component1 
   */
  static test(component) {
    new ComponentSnapshotTester(component).test();
  }
}