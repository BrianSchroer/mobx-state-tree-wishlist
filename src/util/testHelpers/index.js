export { default as mobxSnapshotHelper } from './mobxSnapshotHelper/snapshotHelper';
export { default as ComponentSnapshotHelper } from './ComponentSnapshotHelper/SnapshotHelper';