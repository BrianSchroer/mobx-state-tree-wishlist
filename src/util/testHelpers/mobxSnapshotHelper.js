import { getSnapshot } from 'mobx-state-tree';

export default class {
  static createSnapshot(target) {
    return getSnapshot(target);
  }

  static assertMatch(target) {
    expect(this.createSnapshot(target)).toMatchSnapshot();
  }
}