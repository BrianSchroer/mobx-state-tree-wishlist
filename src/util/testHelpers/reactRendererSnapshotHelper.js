import renderer from 'react-test-renderer';

/**
 * react-test-renderer snapshot helper methods
 */
export default class {

  /**
   * Create component from JSX
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {object} component object
   */
  static createComponent(componentJsx) {
    return renderer.create(componentJsx);
  }

  /**
   * Create component from JSX and serialize as JSON "snapshot" string.
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {string} JSON string.
   */
  static createSnapshot(componentJsx) {
    return this.createComponent(componentJsx).toJSON();
  }

  /**
   * Create component from JSX, serialize as JSON string, and compare to accepted snapshot.
   * If a snapshot doesn't already exist, one is saved. If a previous snapshot exists
   * and the new snapshot doesn't match, you'll be asked if that's an error or if you want
   * to accept the new snapshot.
   * @param {JSX} componentJsx (e.g. '<Button />')
   */
  static assertMatch(componentJsx) {
    expect(this.createSnapshot(componentJsx)).toMatchSnapshot();
  }
}