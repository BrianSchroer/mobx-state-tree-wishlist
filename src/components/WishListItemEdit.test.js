import React from 'react';
import WishListItemEdit from './WishListItemEdit';
import { ComponentSnapshotHelper } from '../util/testHelpers';

const testInput = {
  name: 'Chronicles of Narnia Box Set - C.S. Lewis',
  price: 28.83,
  image: 'https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg'
};

describe('WishListItemEdit', () => {
  const wishListItemEdit = <WishListItemEdit item={testInput} />;

  it('should be a MobXReactObserver', () => {
    expect(wishListItemEdit.type.isMobXReactObserver).toBe(true);
  });

  it('should render correctly', () => {
    ComponentSnapshotHelper.test(wishListItemEdit);
  });
});