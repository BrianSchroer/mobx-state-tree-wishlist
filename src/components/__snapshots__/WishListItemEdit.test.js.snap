// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WishListItemEdit should render correctly 1`] = `
<div
  className="item-edit"
>
  <div>
    <div
      className="item-edit-label"
    >
      <label
        htmlFor="name"
      >
        Thing:
      </label>
    </div>
    <input
      id="name"
      onChange={[Function]}
      value="Chronicles of Narnia Box Set - C.S. Lewis"
    />
  </div>
  <div>
    <div
      className="item-edit-label"
    >
      <label
        htmlFor="price"
      >
        Price:
      </label>
    </div>
    <input
      id="price"
      onChange={[Function]}
      value={28.83}
    />
  </div>
  <div>
    <div
      className="item-edit-label"
    >
      <label
        htmlFor="image"
      >
        Image:
      </label>
    </div>
    <input
      id="image"
      onChange={[Function]}
      value="https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg"
    />
  </div>
</div>
`;
